paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngWriter.kt

```diff
@@ -0,0 +1,276 @@
+import app.cash.paparazzi.internal.apng.PngConstants.Header.ACTL
+import app.cash.paparazzi.internal.apng.PngConstants.Header.FCTL
+import app.cash.paparazzi.internal.apng.PngConstants.Header.FDAT
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IDAT
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IEND
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IHDR
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_BITS_PER_PIXEL
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_COLOR_TYPE_RGB
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_COLOR_TYPE_RGBA
+import okio.Buffer
+import okio.BufferedSink
+import okio.FileSystem
+import okio.Path.Companion.toPath
+import okio.buffer
+import java.awt.Rectangle
+import java.awt.image.BufferedImage
+import java.io.Closeable
+import java.io.File
+import java.lang.IllegalStateException
+import java.lang.Integer.max
+import java.util.zip.CRC32
+import java.util.zip.Deflater
+
+/**
+ * Provides a basic implementation for encoding APNG's.
+ *
+ * Refs: https://www.w3.org/TR/png/#4Concepts.Encoding, https://wiki.mozilla.org/APNG_Specification
+ */
+internal class ApngWriter(
+  private val file: File,
+  private val totalFrameCount: Int,
+  fps: Int
+) : Closeable {
+
+  private val fpsNum = 1
+  private val fpsDen = fps
+  private val tempFile = File(FileSystem.SYSTEM_TEMPORARY_DIRECTORY.toFile(), "temp-${file.name}").apply { deleteOnExit() }
+  private val fileHandle = FileSystem.SYSTEM.openReadWrite(tempFile.path.toPath())
+  private val sink = fileHandle.sink().buffer()
+  private val crcEngine = CRC32()
+
+  private lateinit var firstFrame: BufferedImage
+  private lateinit var previousFrame: BufferedImage
+
+  private var maxWidth = 0
+  private var maxHeight = 0
+  private var ihdrWidth = 0
+  private var ihdrHeight = 0
+  private var iDATChunkEndOffset = 0L
+  private var frameCount = 0
+  private var sequenceNumber = 0
+
+  fun writeImage(image: BufferedImage) {
+    if (frameCount == 0) {
+      sink.writePngSignature()
+      sink.writeIHDR(image)
+
+      if (totalFrameCount > 1) {
+        sink.writeACTL(totalFrameCount, 0)
+        sink.writeFCTL(sequenceNumber++, Rectangle(image.width, image.height))
+      }
+
+      sink.writeIDAT(image)
+      iDATChunkEndOffset = fileHandle.position(sink)
+      firstFrame = image
+    } else {
+      val subRect = ImageUtils.smallestDiffRect(previousFrame, image) ?: MIN_FRAME_RECT
+      sink.writeFCTL(sequenceNumber++, subRect)
+      sink.writeFDAT(sequenceNumber++, subRect, image)
+    }
+
+    maxWidth = max(maxWidth, image.width)
+    maxHeight = max(maxHeight, image.height)
+    previousFrame = image
+    frameCount++
+  }
+
+  override fun close() {
+    if (frameCount != totalFrameCount) {
+      throw IllegalStateException("Expected $totalFrameCount total frames, actual count $frameCount")
+    }
+
+    sink.writeIEND()
+    fileHandle.resize(fileHandle.position(sink))
```

Comment:
Is this necessary?

Reply:
Maybe? Will check, I remember in the case that this file already exists... except looking at it now it's the temp file which we cleanup with deleteOnExit so I'm thinking it is a relic of before I added that in